compiler/rustc_builtin_macros/src/autodiff.rs

```diff
@@ -0,0 +1,707 @@
+fn name(x: &NestedMetaItem) -> String {
+    first_ident(x).name.to_string()
+}
+
+#[cfg(not(llvm_enzyme))]
+pub fn from_ast(
+    ecx: &mut ExtCtxt<'_>,
+    meta_item: &ThinVec<NestedMetaItem>,
+    has_ret: bool,
+) -> AutoDiffAttrs {
+    let mode = name(&meta_item[1]);
+    let mode = match DiffMode::from_str(&mode) {
+        Ok(x) => x,
+        Err(_) => {
+            ecx.sess
+                .dcx()
+                .emit_err(errors::AutoDiffInvalidMode { span: meta_item[1].span(), mode });
+            return AutoDiffAttrs::inactive();
+        }
+    };
+    let mut activities: Vec<DiffActivity> = vec![];
+    for x in &meta_item[2..] {
+        let activity_str = name(&x);
+        let res = DiffActivity::from_str(&activity_str);
+        match res {
+            Ok(x) => activities.push(x),
+            Err(_) => {
+                ecx.sess.dcx().emit_err(errors::AutoDiffUnknownActivity {
+                    span: x.span(),
+                    act: activity_str,
+                });
+                return AutoDiffAttrs::inactive();
+            }
+        };
+    }
+
+    // If a return type exist, we need to split the last activity,
+    // otherwise we return None as placeholder.
+    let (ret_activity, input_activity) = if has_ret {
+        activities.split_last().unwrap()
+    } else {
+        (&DiffActivity::None, activities.as_slice())
+    };
+
+    AutoDiffAttrs { mode, ret_activity: *ret_activity, input_activity: input_activity.to_vec() }
+}
+
+#[cfg(not(llvm_enzyme))]
+pub fn expand(
+    ecx: &mut ExtCtxt<'_>,
+    expand_span: Span,
+    meta_item: &ast::MetaItem,
+    mut item: Annotatable,
+) -> Vec<Annotatable> {
+    // first get the annotable item:
+    let (sig, is_impl): (FnSig, bool) = match &item {
+        Annotatable::Item(ref iitem) => {
+            let sig = match &iitem.kind {
+                ItemKind::Fn(box ast::Fn { sig, .. }) => sig,
+                _ => {
+                    ecx.sess
+                        .dcx()
+                        .emit_err(errors::AutoDiffInvalidApplication { span: item.span() });
+                    return vec![item];
+                }
+            };
+            (sig.clone(), false)
+        }
+        Annotatable::AssocItem(ref assoc_item, _) => {
+            let sig = match &assoc_item.kind {
+                ast::AssocItemKind::Fn(box ast::Fn { sig, .. }) => sig,
+                _ => {
+                    ecx.sess
+                        .dcx()
+                        .emit_err(errors::AutoDiffInvalidApplication { span: item.span() });
+                    return vec![item];
+                }
+            };
+            (sig.clone(), true)
+        }
+        _ => {
+            ecx.sess.dcx().emit_err(errors::AutoDiffInvalidApplication { span: item.span() });
+            return vec![item];
+        }
+    };
+
+    let meta_item_vec: ThinVec<NestedMetaItem> = match meta_item.kind {
+        ast::MetaItemKind::List(ref vec) => vec.clone(),
```

Comment:
yep:
```
error[E0507]: cannot move out of `*sig` which is behind a shared reference
   --> compiler/rustc_builtin_macros/src/autodiff.rs:133:14
    |
133 |             (*sig, false)
    |              ^^^^ move occurs because `*sig` has type `FnSig`, which does not implement the `Copy` trait
    |
help: consider cloning the value if the performance cost is acceptable
    |
133 -             (*sig, false)
133 +             (sig.clone(), false)
    |
```